platforms/core-runtime/daemon-protocol/src/main/java/org/gradle/tooling/internal/provider/serialization/ReplacingObjectOutputStream.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.internal.provider.serialization;
+
+import org.gradle.api.NonNullApi;
+
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.util.Arrays;
+
+@NonNullApi
+public class ReplacingObjectOutputStream extends ObjectOutputStream {
+    private final PayloadSerializer payloadSerializer;
+    private final Class<?>[] classes;
+
+    public ReplacingObjectOutputStream(OutputStream outputStream, PayloadSerializer payloadSerializer, Class<?>... classes) throws IOException {
+        super(outputStream);
+        this.payloadSerializer = payloadSerializer;
+        this.classes = classes;
+        enableReplaceObject(true);
+    }
+
+    @Override
+    protected final Object replaceObject(Object obj) throws IOException {
+        if (isReplaceable(obj)) {
```

Comment:
I assume. only `AdditionalData` is replaceable at the moment.

Reply:
I didn't test with other data types, but anything that is serializable should be replaceable.